test_container/tests/test/pandas/all/dataframe_memory_leak.py

```diff
@@ -0,0 +1,228 @@
+            /
+
+        ''' % (self.col_defs_str, self.col_defs_str, self.num_rows - 1))
+        self.query(udf_def_str)
+        select_sql = 'SELECT foo(%s) FROM FN2.TEST1' % (self.col_names_str)
+        rows = self.query(select_sql)
+        self.assertEqual(self.num_rows, len(rows))
+
+    def test_dataframe_scalar_returns(self):
+        """
+        This test checks that the largest memory block of a tracemalloc snapshot diff is not larger than 100KB, where
+        the memory block snapshots are retrieved during the first/last invocation of the scalar UDF.
+        Reasoning for 100KB is that the number of rows is > 100K, so if there was 1 Byte leaking during every execution,
+        it would be found here.
+        """
+        udf_sql = udf.fixindent('''
+            CREATE OR REPLACE PYTHON3 SCALAR SCRIPT
+            foo(%s)
+            RETURNS DECIMAL(10,5) AS
+
+
+            %%perNodeAndCallInstanceLimit 1;
+
+            import tracemalloc
+            import gc
+            snapshot_begin = None
+            memory_check_executed = False
+            tracemalloc.start()
+            counter = 0
+
+            def run(ctx):
+                df = ctx.get_dataframe()
+                global memory_check_executed
+                global snapshot_begin
+                global counter
+                if counter == 0:
+                    print("Retrieving start snapshot", flush=True)
+                    snapshot_begin = tracemalloc.take_snapshot()
+                if counter == %s:
+                    assert memory_check_executed == False #Sanity check for row number
+                    print("Checking memory usage", flush=True)
+                    gc.collect()
+                    snapshot_end = tracemalloc.take_snapshot()
+                    top_stats_begin_end = snapshot_end.compare_to(snapshot_begin, 'lineno')
+                    first_item = top_stats_begin_end[0] #First item is always the largest one
+                    if first_item.size_diff > 100000:
+                        raise RuntimeError(f"scalar emit UDF uses too much memory: {first_item}")
+                    memory_check_executed = True
+                counter = counter + 1
+
+                return (df.iloc[0, 0] + df.iloc[0, 1]).item()
+            /
+            ''' % (self.col_defs_str, self.num_rows - 1))
+        self.query(udf_sql)
+        print(udf_sql)
+        select_sql = 'SELECT foo(%s) FROM FN2.TEST1' % (self.col_names_str)
+        print(select_sql)
+        rows = self.query(select_sql)
+        self.assertEqual(self.num_rows, len(rows))
+
+
+    def test_dataframe_set_emits(self):
+        """
+        This test validates that the <EXASCRIPT> module does not leak more than 100kb of RAM in a set/emits UDF.
+        The test is different from the others as it checks only the <EXASCRIPT> for leaks,
+        not the rest of the UDF client; the reason for that is that a
+        set UDF reads all input rows at once into a dataframe.
```

Comment:
but you could also read and emit batches and that require multiple call to get_dataframe and each call could leak for its self.

Reply:
Done. Pls check